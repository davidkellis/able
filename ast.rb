class PackageDefinition
  attr_accessor :name, :expressions

  def initialize(name, expressions)
    @name = name
    @expressions = expressions
  end

  def imports
    @imports ||= expressions.select {|e| e.is_a? Import }
  end
end

class Import
  attr_accessor :package_name, :package_alias, :identifiers

  # identifiers is either an array of string identifiers, or a map of string => string identifiers and corresponding aliases
  def initialize(package_name, package_alias, identifiers)
    @package_name, @package_alias, @identifiers = package_name, package_alias, identifiers
  end
end

class FunctionDefinition
  attr_accessor :type_parameters, :function_name, :parameters, :return_type, :expressions

  def initialize(type_parameters, function_name, parameters, return_type, expressions)
    @type_parameters, @function_name, @parameters, @return_type, @expressions = type_parameters, function_name, parameters, return_type, expressions
  end
end

class FunctionCall
  attr_accessor :function_name, :arguments

  def initialize(function_name, arguments)
    @function_name, @arguments = function_name, arguments
  end
end

class InferredType
end

class StringLiteral
  attr_accessor :value

  def initialize(value)
    @type = type
    @value = value
  end
end

class Identifier
  attr_accessor :name

  def initialize(name)
    @name = name
  end
end

class RubyEmitter
  attr_accessor :ast, :out

  def emit(ast, out = $stdout)
    @ast, @out = ast, out

    case ast
    when PackageDefinition
      build_package_definition(ast)
    end
  end

  def emit_package_definition(package_definition)
    buffer = <<-EOF
package #{package_definition.name}

#{emit_import()}
    EOF
  end

  def emit_import(import, indent_level = 0)
  end
end


class GoEmitter
  attr_accessor :ast, :out

  def emit(ast, out = $stdout)
    @ast, @out = ast, out

    case ast
    when PackageDefinition
      build_package_definition(ast)
    end
  end

  def emit_package_definition(package_definition)
    buffer = <<-EOF
package #{package_definition.name}

#{emit_import()}
    EOF
  end

  def emit_import(import, indent_level = 0)
  end
end


def hello_world_ast
# package main
# import io.puts
# fn main() {
#   puts("hello world")
# } keyb

  ast = PackageDefinition.new("main", [
    Import.new("io", nil, ["puts"]),
    FunctionDefinition.new("main", [], InferredType.new, [
      FunctionCall.new("puts", [
        Literal.new("hello world")
      ])
    ])
  ])
  emitter = GoEmitter.new
  emitter.emit(ast)
end

hello_world_ast
