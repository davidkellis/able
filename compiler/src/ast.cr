require "arborist"

module Ast
  class Node
    property parent : Node?
    
    # children returns the list of all actual/literal child Node nodes in AST tree structure, 
    def children : Array(Node)
      raise "Node#children is an abstract method"
    end

    def child : Node
      children.first
    end

    def child? : Node?
      children.first?
    end

    # returns all descendants listed out in a pre-order traversal
    def descendants : Array(Node)
      nodes = [] of Node
      accumulate = ->(node : Node) { nodes << node }
      children.each {|child| child.preorder_traverse(accumulate) }
      nodes
    end

    def s_exp(indent : Int32 = 0) : String
      raise "Node#s_exp is an abstract method"
    end

    def preorder_traverse(visit : Node -> _)
      visit.call(self)
      children.each {|child| child.preorder_traverse(visit) }
    end

    def postorder_traverse(&visit : Node -> _)
      postorder_traverse(visit)
    end

    def postorder_traverse(visit : Node -> _)
      children.each {|child| child.postorder_traverse(visit) }
      visit.call(self)
    end

    def root?
      @parent.nil?
    end

    # returns all nodes listed out in a pre-order traversal
    def self_and_descendants : Array(Node)
      ([self] of Node).concat(descendants)
    end

  end

  class File < Node
    @package_declaration : PackageDecl?
    property package_level_declarations : Array(ImportDecl | AssignmentExpr)

    def initialize
      @package_level_declarations = Array(ImportDecl | AssignmentExpr).new
    end

    def package_declaration : PackageDecl
      @package_declaration || raise "File#package_declaration undefined"
    end

    def package_declaration=(package_decl : PackageDecl)
      @package_declaration = package_decl
      self
    end

    def add_package_level_declaration(package_level_declaration : Node)
      case package_level_declaration
      when ImportDecl
        @package_level_declarations.push(package_level_declaration)
      when AssignmentExpr
        @package_level_declarations.push(package_level_declaration)
      else
        raise "Unexpected package level declaration: #{typeof(package_level_declaration)}"
      end
    end

    def children : Array(Node)
      [package_declaration].concat(@package_level_declarations)
    end

    def s_exp(indent : Int32 = 0) : String
      prefix = " " * indent
      package_decl_sexp = package_declaration.s_exp(indent+2)
      package_level_decl_sexp = package_level_declarations.map(&.s_exp(indent+2)).join("\n")
      "#{prefix}(file\n#{package_decl_sexp}\n#{package_level_decl_sexp})"
    end
  end

  class PackageDecl < Node
    @name : String?

    def name : String
      @name || raise "PackageDecl#name undefined"
    end

    def name=(name : String)
      @name = name
      self
    end

    def s_exp(indent : Int32 = 0) : String
      prefix = " " * indent
      "#{prefix}(package_decl #{name})"
    end
  end

  class ImportDecl < Node
    getter type : Symbol    # :all (import all members from package), :some (import specified members from package), :package (import only package name)
    getter package_name : String    # a fully-qualified package name
    getter package_alias : String?  # a single-component package name
    getter members : Hash(String, String)  # a list of members to import from `@package_name`; the key is the member name to import, the value is the alias (in the case that no member alias is defined, the member name is the alias)

    def self.all(fq_package_name : String, package_alias : String?)
      new(:all, fq_package_name, package_alias, Hash(String, String).new)
    end

    def self.some(fq_package_name : String, package_alias : String?, members : Hash(String, String))
      new(:some, fq_package_name, package_alias, members)
    end

    def self.package(fq_package_name : String, package_alias : String?)
      new(:package, fq_package_name, package_alias, Hash(String, String).new)
    end

    def initialize(@type, @package_name, @package_alias, @members)
    end

    def s_exp(indent : Int32 = 0) : String
      prefix = " " * indent
      "#{prefix}(import_decl #{@type} #{@package_name} #{@package_alias || "nil"} [#{@members.map{|k,v| "#{k}=>#{v}" }.join(",")}])"
    end
  end

  class AssignmentExpr < Node
  end

  def self.build(parse_tree : Arborist::ParseTree) : Node
    matcher = Arborist::Matcher.new

    visitor = Arborist::Visitor(Node).new

    visitor.on("file") do |node|
      file_ast = File.new
      file_ast.package_declaration = node.capture("PackageDecl").visit(visitor).as(PackageDecl)
      # file_ast.package_level_declarations = node.captures("PackageLevelDecl").map(&.visit(visitor).as(ImportDecl | AssignmentExpr))
      node.captures("PackageLevelDecl").each do |node|
        # todo: fix this; we want to visit the PackageLevelDecl node, which should return an ImportDecl node or a AssignmentExpr node, which we then want to add to the 
        package_level_decl_node = node.visit(visitor)
        file_ast.add_package_level_declaration(package_level_decl_node)
      end
      file_ast
    end

    visitor.on("PackageDecl") do |node|
      package_decl_ast = PackageDecl.new
      package_decl_ast.name = node.capture("package_ident").text
      package_decl_ast
    end

    visitor.on("PackageLevelDecl_import_decl") do |node|
      puts "*" * 80
      node.capture("import_decl").visit(visitor)
      # import_decl_seq_node = node.capture("import_decl")
      # import_decl_apply_node = import_decl_seq_node.capture("import_decl")
      # import_decl_apply_node.visit(visitor)
    end

    visitor.on("import_decl_all") do |node|
      puts "1" * 80
      package_name = node.capture("package_name").text
      package_alias = node.capture?("package_alias").try(&.text)
      ImportDecl.all(package_name, package_alias)
    end

    visitor.on("import_decl_some") do |node|
      puts "2" * 80
      package_name = node.capture("package_name").text
      package_alias = node.capture?("package_alias").try(&.text)
      members : Hash(String,String) = node.captures("members").map do |member_node|
        member_name = member_node.capture("member_name").text
        member_alias = member_node.capture?("member_alias").try(&.text) || member_name    # if no member alias is defined, then we define the alias to be the original member name
        [member_name, member_alias]
      end.to_h
      ImportDecl.some(package_name, package_alias, members)
    end

    visitor.on("import_decl_package") do |node|
      puts "3" * 80
      package_name = node.capture("package_name").text
      package_alias = node.capture?("package_alias").try(&.text)
      ImportDecl.package(package_name, package_alias)
    end

    visitor.visit(parse_tree)
  end
end