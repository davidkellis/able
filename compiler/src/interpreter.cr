class Interpreter
  module ObjectInterface
    def inspect : String
      "Object{id=#{object_id}}"
    end

    def type : String
      raise "Object#type not implemented"
    end
  end

  class Bindings
    property @bindings : Hash(string, Object)
    
    def initialize()
      @bindings = Hash(string, Object).new
    end

    def add(name, value)
      @bindings[name] = value
    end

    def remove(name)
      @bindings.delete(name)
    end
  end

  alias Object = Bool | U8 | I8 | I32 | I64 | F32 | F64 | Bool | String | Function

  class Bool
    include ObjectInterface

    property value : ::Bool

    def initialize(@value)
    end

    def inspect
      "#{@value}"
    end

    def type : String
      "bool"
    end
  end

  class Number(T)
    include ObjectInterface

    property value : T

    def initialize(@value)
    end

    def inspect
      "#{@value}"
    end
  end

  class Int(T) < Number(T)
  end

  class Float(T) < Number(T)
  end

  class U8 < Int(::UInt8)
    def type
      "u8"
    end
  end

  class I8 < Int(::Int8)
    def type
      "i8"
    end
  end

  class I32 < Int(::Int32)
    def type
      "i32"
    end
  end

  class I64 < Int(::Int64)
    def type
      "i64"
    end
  end

  class F32 < Float(::Float32)
    def type
      "f32"
    end
  end

  class F64 < Float(::Float64)
    def type
      "f64"
    end
  end

  class String
    include ObjectInterface

    property value : ::String

    def initialize(@value)
    end

    def inspect
      "\"#{@value}\""
    end

    def type : String
      "string"
    end
  end

  class Block
    property expressions : Array(Ast::Expression)

    def initialize(@expressions)
    end
  end

  class Function
    include ObjectInterface

    property parameter_names : Array(string)
    property bindings : Bindings
    property body : Block

    def initialize(@parameter_names, @bindings, @body)
    end

    def inspect
      "\"#{@value}\""
    end

    def type : String
      "string"
    end
  end


  property global_bindings : Bindings

  def initialize
    @global_bindings = Bindings.new

  end

  def eval(node : Ast::Node) : Object
    case node
    when Ast::Package
    when Ast::Integer
    when Ast::Float
    when Ast::Branch
    when Ast::Loop
    when Ast::Return
    when Ast::Break
      
    end
  end
end
