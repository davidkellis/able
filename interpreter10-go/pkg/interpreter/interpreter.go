package interpreter

import (
	"fmt"
	"strings"

	"able/interpreter10-go/pkg/ast"
	"able/interpreter10-go/pkg/runtime"
)

type packageMeta struct {
	namePath  []string
	isPrivate bool
}

// Interpreter drives evaluation of Able v10 AST nodes.
type Interpreter struct {
	global          *runtime.Environment
	inherentMethods map[string]map[string]*runtime.FunctionValue
	interfaces      map[string]*runtime.InterfaceDefinitionValue
	implMethods     map[string][]implEntry
	unnamedImpls    map[string]map[string]map[string]struct{}
	raiseStack      []runtime.Value
	packageRegistry map[string]map[string]runtime.Value
	packageMetadata map[string]packageMeta
	currentPackage  string
	breakpoints     []string
}

func identifiersToStrings(ids []*ast.Identifier) []string {
	parts := make([]string, 0, len(ids))
	for _, id := range ids {
		if id == nil {
			continue
		}
		parts = append(parts, id.Name)
	}
	return parts
}

func joinIdentifierNames(ids []*ast.Identifier) string {
	parts := identifiersToStrings(ids)
	if len(parts) == 0 {
		return ""
	}
	return strings.Join(parts, ".")
}

func (i *Interpreter) qualifiedName(name string) string {
	if i.currentPackage == "" {
		return ""
	}
	return i.currentPackage + "." + name
}

func (i *Interpreter) pushBreakpoint(label string) {
	i.breakpoints = append(i.breakpoints, label)
}

func (i *Interpreter) popBreakpoint() {
	if len(i.breakpoints) == 0 {
		return
	}
	i.breakpoints = i.breakpoints[:len(i.breakpoints)-1]
}

func (i *Interpreter) hasBreakpoint(label string) bool {
	for idx := len(i.breakpoints) - 1; idx >= 0; idx-- {
		if i.breakpoints[idx] == label {
			return true
		}
	}
	return false
}

func (i *Interpreter) registerSymbol(name string, value runtime.Value) {
	if i.currentPackage == "" {
		return
	}
	bucket, ok := i.packageRegistry[i.currentPackage]
	if !ok {
		bucket = make(map[string]runtime.Value)
		i.packageRegistry[i.currentPackage] = bucket
	}
	bucket[name] = value
	if qn := i.qualifiedName(name); qn != "" {
		i.global.Define(qn, value)
	}
}

// New returns an interpreter with an empty global environment.
func New() *Interpreter {
	return &Interpreter{
		global:          runtime.NewEnvironment(nil),
		inherentMethods: make(map[string]map[string]*runtime.FunctionValue),
		interfaces:      make(map[string]*runtime.InterfaceDefinitionValue),
		implMethods:     make(map[string][]implEntry),
		unnamedImpls:    make(map[string]map[string]map[string]struct{}),
		raiseStack:      make([]runtime.Value, 0),
		packageRegistry: make(map[string]map[string]runtime.Value),
		packageMetadata: make(map[string]packageMeta),
		breakpoints:     make([]string, 0),
	}
}

// GlobalEnvironment returns the interpreterâ€™s global environment.
func (i *Interpreter) GlobalEnvironment() *runtime.Environment {
	return i.global
}

// EvaluateModule executes a module node and returns the last evaluated value and environment.
func (i *Interpreter) EvaluateModule(module *ast.Module) (runtime.Value, *runtime.Environment, error) {
	moduleEnv := i.global
	prevPackage := i.currentPackage
	defer func() { i.currentPackage = prevPackage }()

	if module.Package != nil {
		moduleEnv = runtime.NewEnvironment(i.global)
		pkgParts := identifiersToStrings(module.Package.NamePath)
		pkgName := strings.Join(pkgParts, ".")
		i.currentPackage = pkgName
		if _, ok := i.packageRegistry[pkgName]; !ok {
			i.packageRegistry[pkgName] = make(map[string]runtime.Value)
		}
		i.packageMetadata[pkgName] = packageMeta{
			namePath:  pkgParts,
			isPrivate: module.Package.IsPrivate,
		}
	} else {
		i.currentPackage = ""
	}

	for _, imp := range module.Imports {
		if _, err := i.evaluateImportStatement(imp, moduleEnv); err != nil {
			return nil, nil, err
		}
	}

	var last runtime.Value = runtime.NilValue{}
	for _, stmt := range module.Body {
		val, err := i.evaluateStatement(stmt, moduleEnv)
		if err != nil {
			if rs, ok := err.(raiseSignal); ok {
				return nil, moduleEnv, rs
			}
			if _, ok := err.(returnSignal); ok {
				return nil, nil, fmt.Errorf("return outside function")
			}
			return nil, nil, err
		}
		last = val
	}
	return last, moduleEnv, nil
}

func (i *Interpreter) getPackageMeta(pkgName string, namePath []string) packageMeta {
	if meta, ok := i.packageMetadata[pkgName]; ok {
		return meta
	}
	dup := make([]string, len(namePath))
	copy(dup, namePath)
	return packageMeta{namePath: dup, isPrivate: false}
}
