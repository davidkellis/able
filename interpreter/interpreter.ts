import * as AST from './ast'; // Import our AST definitions

// --- Runtime Values ---
// More closely aligned with spec types
type AblePrimitive =
  | { kind: "i8"; value: number }
  | { kind: "i16"; value: number }
  | { kind: "i32"; value: number }
  | { kind: "i64"; value: bigint }
  | { kind: "i128"; value: bigint }
  | { kind: "u8"; value: number }
  | { kind: "u16"; value: number }
  | { kind: "u32"; value: number }
  | { kind: "u64"; value: bigint }
  | { kind: "u128"; value: bigint }
  | { kind: "f32"; value: number }
  | { kind: "f64"; value: number }
  | { kind: "string"; value: string }
  | { kind: "bool"; value: boolean }
  | { kind: "char"; value: string } // Single Unicode char as string
  | { kind: "nil"; value: null }
  | { kind: "void"; value: undefined }; // Represent void

// Represents a runtime function (closure)
interface AbleFunction {
    node: AST.FunctionDefinition;
    closureEnv: Environment; // Environment captured at definition time
}

// Represents a runtime struct definition
interface AbleStructDefinition {
    kind: 'struct_definition';
    name: string;
    definitionNode: AST.StructDefinition;
    // Store generic info if needed later
}

// Represents a runtime struct instance
interface AbleStructInstance {
    kind: 'struct_instance';
    definition: AbleStructDefinition;
    // Use array for positional, map for named
    values: AbleValue[] | Map<string, AbleValue>;
}

// Represents a runtime union definition (placeholder)
interface AbleUnionDefinition {
    kind: 'union_definition';
    name: string;
    definitionNode: AST.UnionDefinition;
}

// Represents a runtime interface definition (placeholder)
interface AbleInterfaceDefinition {
    kind: 'interface_definition';
    name: string;
    definitionNode: AST.InterfaceDefinition;
}

// Represents a runtime implementation definition (placeholder)
interface AbleImplementationDefinition {
    kind: 'implementation_definition';
    implNode: AST.ImplementationDefinition;
    // Link to interface and target type info
}

// Represents a runtime methods definition (placeholder)
interface AbleMethodsDefinition {
    kind: 'methods_definition';
    methodsNode: AST.MethodsDefinition;
    // Link to target type info
}

// Represents a runtime error value (for !, rescue)
interface AbleError {
    kind: 'error';
    // Based on spec's Error interface concept
    message: string;
    // Add cause, stack trace etc. later
    originalValue?: any; // The value raised
}

// Represents a runtime Proc handle (placeholder)
interface AbleProcHandle {
    kind: 'proc_handle';
    id: number; // Example ID
    // Add status, result promise/callback etc.
}

// Represents a runtime Thunk (placeholder)
interface AbleThunk {
    kind: 'thunk';
    id: number; // Example ID
    // Add logic for lazy evaluation and blocking
}

// Represents a runtime Array (placeholder)
interface AbleArray {
    kind: 'array';
    elements: AbleValue[];
}

// Represents a runtime Range (placeholder)
interface AbleRange {
    kind: 'range';
    start: number | bigint;
    end: number | bigint;
    inclusive: boolean;
}


type AbleValue =
  | AblePrimitive
  | { kind: "function"; value: AbleFunction }
  | AbleStructDefinition
  | AbleStructInstance
  | AbleUnionDefinition // Added
  | AbleInterfaceDefinition // Added
  | AbleImplementationDefinition // Added
  | AbleMethodsDefinition // Added
  | AbleError // Added
  | AbleProcHandle // Added
  | AbleThunk // Added
  | AbleArray // Added
  | AbleRange // Added
  ;

// Special object to signal a `return` occurred
class ReturnSignal {
    constructor(public value: AbleValue) {}
}
// Special object to signal a `raise` occurred
class RaiseSignal {
    constructor(public value: AbleValue) {}
}

