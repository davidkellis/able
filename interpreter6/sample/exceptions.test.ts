import { describe, it, expect, vi } from "vitest";
import { interpret } from "../interpreter";
import exceptionsModule from "./exceptions"; // Import the AST module
import * as AST from "../ast";

// Helper to capture console output (including errors)
function captureOutput(fn: () => void): { stdout: string; stderr: string } {
  const stdout: string[] = [];
  const stderr: string[] = [];
  const logSpy = vi.spyOn(console, "log").mockImplementation((...args) => {
    stdout.push(args.map(String).join(" "));
  });
  const errorSpy = vi.spyOn(console, "error").mockImplementation((...args) => {
    stderr.push(args.map(String).join(" "));
  });

  try {
    fn();
  } finally {
    logSpy.mockRestore();
    errorSpy.mockRestore();
  }
  return { stdout: stdout.join("\n"), stderr: stderr.join("\n") };
}

// Descriptive name based on the original module name
describe("Interpreter Sample - exceptions", () => {
  it("should produce the expected output (including errors)", () => {
    // Exceptions might print to console.error, so capture both
    const output = captureOutput(() => {
      // Cast might be needed
      interpret(exceptionsModule as AST.Module);
    });
    // Snapshot both stdout and stderr
    expect(output.stdout).toMatchSnapshot("stdout");
    expect(output.stderr).toMatchSnapshot("stderr");
  });
});
