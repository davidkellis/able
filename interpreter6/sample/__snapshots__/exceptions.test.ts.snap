// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Interpreter Sample - exceptions > should produce the expected output (including errors) > stderr 1`] = `""`;

exports[`Interpreter Sample - exceptions > should produce the expected output (including errors) > stdout 1`] = `
"--- Running main function ---

--- Exception Handling Tests ---

Test Case 1: 10 / 2
Result 1: 5

Test Case 2: 5 / 0
Result 2: Rescued successfully: Division by zero!

Test Case 3: raise string
Result 3: Caught generic error: A manually raised string error

Test Case 4: Unrescued 1 / 0 (will terminate if uncommented)
--- main function finished ---"
`;

exports[`Interpreter Sample - exceptions should produce the expected output (including errors): stdout 1`] = `
"--- Running main function ---

--- Exception Handling Tests ---

Test Case 1: 10 / 2
Result 1: 5

Test Case 2: 5 / 0
Result 2: Rescued successfully: Division by zero!

Test Case 3: raise string
Result 3: Caught generic error: A manually raised string error

Test Case 4: Unrescued 1 / 0 (will terminate if uncommented)
--- main function finished ---"
`;

exports[`Interpreter Sample - exceptions should produce the expected output (including errors): stderr 1`] = `""`;
